rfcs/1111-first-class-virtual-device.md

```diff
@@ -0,0 +1,146 @@
+
+However, let's say you want to write a pass that uses the virtual devices after device planning. Now, you'll be able to use the virtual device directly in your pass, just like you can with the checked_type_ field.
+
+For example, in this visitor, we can just look at the virtual device directly. 
+
+```
+  Expr VisitExpr_(const LetNode* let_node) final {
+    Expr expr = GetRef<Expr>(let_node);
+    // Iterate through chained lets, provided they all agree on their device type.
+    SEScope scope = expr->virtual_device_;
+    ...
+ ```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The AST change will occur in include/tvm/ir/expr.h:
+
+```
+class RelayExprNode : public BaseExprNode {
+ public:
+  /*!
+   * \brief Stores the result of type inference(type checking).
+   *
+   * \note This can be undefined before type inference.
+   *       This value is discarded during serialization.
+   */
+  mutable Type checked_type_ = Type(nullptr);
+  /*!
+   * \return The checked_type
+   */
+  inline const Type& checked_type() const;
+  /*!
+   * \brief Check if the inferred(checked) type of the Expr
+   *  is backed by a TTypeNode and return it.
+   *
+   * \note This function will thrown an error if the node type
+   *       of this Expr is not TTypeNode.
+   *
+   * \return The corresponding TTypeNode pointer.
+   * \tparam The specific TypeNode we look for.
+   */
+  template <typename TTypeNode>
+  inline const TTypeNode* type_as() const;
+
+  /*!
+   * \brief The virtual device (SEScope) for this node (the result of device planning).
+   *
+   * \note Unfortunately, the type of virtual_device_ needs to be ObjectRef to avoid a circular import.
+   *       We can forward-declare the SEScope type for the getter function, but not for the field
+   *       itself.
+   */
+  mutable ObjectRef virtual_device_;
+
+  /*!
+   * \return The virtual device (currently called SEScope, this will be changing soon.)
+   */
+  SEScope virtual_device() const;
+
+  static constexpr const char* _type_key = "RelayExpr";
+  static constexpr const uint32_t _type_child_slots = 22;
+  TVM_DECLARE_BASE_OBJECT_INFO(RelayExprNode, BaseExprNode);
+};
+```
+
+Additionally, I will add virtual_device_ to the WithFields methods.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+One challenge with making virtual devices first class is that passes in TVM do not propagate all fields when they visit expressions. You can see this today with spans-- most visitors do not preserve spans. When we introduce the virtual device field, we will need to ensure that it is propagated correctly throughout the Relay program. To do this, I introduced WithFields (code: https://github.com/apache/tvm/blob/main/src/relay/ir/expr.cc#L79-L99), a COW constructor that copies extra fields, including spans.  I will extend WithFields to also copy virtual devices. Then, we can use WithFields to ensure that the virtual device field is correctly propagated. 
```

Comment:
Sorry that I didn't follow `WithFields` before but this is indeed a super useful feature exactly due to the reason you pointed out. Could you clarify that whether we need to change all Relay passes to use `WithFields` to preserve the fields, or we could have an approach to make it effective to existing passes without changing them?

Reply:
Thanks the feedback @comaniac! I outlined a few other options we considered in the rationales and alternatives section. I'll expand a bit on what I said there. 

The current approach device planning takes is to store the virtual device (SEScope) information in the Relay program in the on_device nodes. Passes will always preserve these nodes and provided that passes don't significantly change the lexical scope, they won't cause problems. 

If we store the virtual device information directly on nodes, we will need to preserve it, either manually or through the WithFields method. In general passes *should* be doing this anyways-- they shouldn't be throwing away information willy-nilly. In reality they don't, which why spans are pretty much useless right now. 

We need to balance adding overhead to passes with maintaining a global invariant. Based on our experience with the current implementation, we think that the overhead is necessary to maintain the invariant.

Fortunately, the change is only difficult because we have to do it in so many places, but it will be a big PITA, which is why @mbs-octoml used on_device to represent this information in the first place. You can take a look at this PR for an example of what the changes would look like: https://github.com/apache/tvm/pull/9533

Reply:
Fair enough to me about the concern of invariant. I guess we may not need to propagate virtual device after each pass but just need to do it before a pass that requires such information, as long as they are available at the node level (i.e., `WithField` is used everywhere).

I totally agree with you that `WithField` should be used in all passes. I'm just thinking how could we more or less enforce it (e.g., fail to compile TVM otherwise or something like that). Otherwise it's hard to make sure all pass implementations in the future would follow this practice.

Reply:
Hi Cody, thanks so much for the comments. In my travels with PlanDevices I've seen four flavors of passes.
 a) Those that just want to know the device for every sub-expression on demand (eg manifest memory). Currently those passes need the help of the DeviceAware* visitors, but with this change the device will be right at hand.
 b) Those that rewrite expressions fairly deeply and should be fully device aware ones (eg lambda lifting). Currently those accumulate and re-encode devices with the existing on_device+function attributes convention. With this change they'll do the same, but the bookkeeping should be very much simpler.
 c) Those that can be completely ignorant of devices and just do local rewriting (eg a pattern based local optimization). The before and after expressions should remain on the same device. Currently the lexical scoping rules for on_device annotations lets us avoid doing anything special for these passes -- we don't even have to worry about re-running PlanDevices. But with this change we'll not be able to get away with that anymore and we'll need to rerun PlanDevices (hopefully in an efficient form) to 'fill in the missing devices'.
 d) Passes which may move expressions around (eg dead code elimination when inlining is on, constant folding), but those expressions just need to bring their devices along with them. Currently these passes have proven to be quite fragile and the source of some nasty bugs. With this change I think they will become much more robust.

So this change is intended to help with d), but in return we need to deal with re-running PlanDevices for c).

I think passes which respectfully decline to use WithFields are just a special case of c). I think (but I'm prepared to be surprised) PlanDevices should have an 'erasure' property: ie
 PlanDevices(EraseAllVirtualDevices(PlanDevices(x)) = PlanDevices(x)
So those passes will incur the expense of reconstructing devices from scratch until they help out by switching to WithFields, but otherwise are semantically sound. Ie WithFields can be 'opt in' rather than required.

Anyway that was my rough thinking, but I remain a bit worried about the 'erasure' property so perhaps more passes than we like will need to be moved out of c) into b) or d).

Reply:
Oh, one more thing: For completeness we should mention there's a third approach (neither the current impl nor this RFC). PlanDevices inserts device_copy where needed, but otherwise doesn't bother trying to make it easy for downstream passes to recover the device for every sub-expression. Instead we have a helper visitor that just constructs that as an Map<Expr, SEScope>, and any passes that need it just call the helper. That's roughly how things used to work before I went down the current path, and maybe we'll need to go back to it should this RFC prove intractable. But I'm pretty confident we can do better and handle the c) style passes cheaply.

Reply:
Thanks for the detail analysis and explanation! I agree with your conclusions and feel conformable with the plan now. I'm now thinking as long as we could have unit tests running through the passes, we should be able to catch all missing information.